proposals/0056-route-caching.md

```diff
@@ -0,0 +1,1149 @@
+
+### Invalidation Implementation
+
+Each cache provider implements invalidation differently:
+
+**Vercel Provider:**
+
+```ts
+// Call Vercel's edge cache invalidation API
+// https://vercel.com/docs/rest-api/reference/endpoints/edge-cache/invalidate-by-tag
+await fetch(`https://api.vercel.com/v1/edge-cache/invalidate-by-tags`, {
+  method: "POST",
+  headers: {
+    Authorization: `Bearer ${process.env.VERCEL_TOKEN}`,
+    "Content-Type": "application/json",
+  },
+  body: JSON.stringify({
+    tags: [tag],
+    target: "production",
+  }),
+});
+```
+
+**Netlify Provider:**
+
+```ts
+// Use Netlify's purgeCache helper
+// https://docs.netlify.com/build/caching/caching-overview/#purge-by-cache-tag
+import { purgeCache } from "@netlify/functions";
+
+await purgeCache({ tags: [tag] });
+```
+
+**Cloudflare Workers Provider:**
+
+```ts
+// Look up tag mappings in KV
+const routes = await env.CACHE_TAGS.get(`tag:${tag}`, { type: "json" });
+// Delete each route from cache
+for (const route of routes) {
+  await caches.default.delete(route);
+}
+```
+
+**Cloudflare CDN Provider:**
+
+```ts
+// Call Cloudflare's purge API
+// https://developers.cloudflare.com/api/resources/cache/methods/purge/
+await fetch(
+  `https://api.cloudflare.com/client/v4/zones/${zoneId}/purge_cache`,
+  {
+    method: "POST",
+    headers: {
+      Authorization: `Bearer ${process.env.CF_TOKEN}`,
+      "Content-Type": "application/json",
+    },
+    body: JSON.stringify({ tags: [tag] }),
+  }
+);
+```
+
+**Node Memory Provider:**
+
+```ts
+// Clear from in-memory cache
+cache.delete(key);
+// Or clear all entries with a tag
+for (const [key, value] of cache.entries()) {
+  if (value.tags?.includes(tag)) {
+    cache.delete(key);
+  }
+}
+```
+
+### Adapter Integration
+
+Adapters provide default cache drivers via the integration API, similar to how they provide session drivers. Users can override the default driver in their config.
+
+**Driver Loading:**
+
+Similar to sessions, cache drivers are loaded at runtime. Driver packages are referenced by their module ID, and dynamically imported. The Netlify, Vercel, Cloudflare Workers, and Node memory drivers are published as subpath exports of their respective adapter packages.
+
+**Adapter providing a default:**
+
+```ts
+// Example: @astrojs/vercel adapter
+export function vercel() {
+  return {
+    name: "@astrojs/vercel",
+    hooks: {
+      "astro:config:done": ({ setAdapter }) => {
+        setAdapter({
+          name: "@astrojs/vercel",
+          // ...other adapter properties
+          // Provide default cache driver
+          cache: {
+            driver: "@astrojs/vercel/cache", // Default
+          },
+        });
+      },
+    },
+  };
+}
+```
+
+**User overriding the default with a driver package:**
+
+```ts
+// astro.config.ts
+export default defineConfig({
+  adapter: node(),
+  cache: {
+    // Override Node's default in-memory cache with Fastly
+    driver: "@astrojs/cache-fastly",
```

Comment:
I think I'd prefer an API like fonts, where you call a function. That would make it easier to have custom drivers:
```js
driver: cacheFastly({
	serviceId: process.env.FASTLY_SERVICE_ID,
	token: process.env.FASTLY_TOKEN,
})
```

Reply:
Ah got it. Yeah, that could be a better pattern, as it makes it easier to type the options

Reply:
LGTM!